File: ResNet50_predict.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/dnn.hpp>
#include <fstream>
#include <iostream>
#include <vector>
#include <chrono>
#include <algorithm>

using namespace cv;
using namespace cv::dnn;
using namespace std;

vector<string> loadClassNames(const string& filePath) {
    vector<string> classNames;
    ifstream file(filePath);

    if (!file.is_open()) {
        cerr << "Error: Could not open class names file: " << filePath << endl;
        return classNames;
    }

    string line;
    while (getline(file, line)) {
        if (!line.empty()) {
            classNames.push_back(line);
        }
    }

    file.close();

    if (classNames.size() != 90) {
        cerr << "Warning: Expected 90 classes, but got " << classNames.size() << endl;
    }

    return classNames;
}

Mat preprocessImage(const string& imagePath) {
    Mat img = imread(imagePath, IMREAD_COLOR); // 确保读取为彩色图像
    if (img.empty()) {
        cerr << "Error: Could not read image: " << imagePath << endl;
        return Mat();
    }

    Mat imgResized;
    resize(img, imgResized, Size(224, 224));

    // 检查通道数
    if (imgResized.channels() != 3) {
        cerr << "Error: Processed image does not have 3 channels" << endl;
        return Mat();
    }


    cvtColor(imgResized, imgResized, COLOR_BGR2RGB);
    
    return imgResized;
}

void predictImage(const string& imagePath, Net& model, const vector<string>& classNames) {
    Mat img = preprocessImage(imagePath);
    if (img.empty()) {
        cerr << "Error: Failed to load or preprocess image." << endl;
        return;
    }

    // 检查图像通道数
    if (img.channels() != 3) {
        cerr << "Error: Image does not have 3 channels. Actual channels: " << img.channels() << endl;
        return;
    }

    auto start = chrono::high_resolution_clock::now();

    // 创建blob并进行归一化（1/255）
    Mat imgRGB;
    cvtColor(img, imgRGB, COLOR_BGR2RGB);  // 显式转换颜色空间
    Mat blob = blobFromImage(imgRGB, 1.0 / 255.0, Size(224, 224), Scalar(0, 0, 0), true, false);

    // 检查blob维度
    if (blob.dims != 4 || blob.size[1] != 3) {
        cerr << "Error: Blob dimensions incorrect. Expected [1,3,224,224], got ["
            << blob.size[0] << "," << blob.size[1] << "," << blob.size[2] << "," << blob.size[3] << "]" << endl;
        return;
    }

    // 标准化（减去均值，除以标准差）
    float* blobData = blob.ptr<float>();
    int totalPixels = 224 * 224;
    const float mean[] = { 0.485f, 0.456f, 0.406f };
    const float std[] = { 0.229f, 0.224f, 0.225f };

    for (int c = 0; c < 3; ++c) {
        for (int i = 0; i < totalPixels; ++i) {
            blobData[c * totalPixels + i] = (blobData[c * totalPixels + i] - mean[c]) / std[c];
        }
    }

    // 设置输入并前向传播
    model.setInput(blob);
    Mat outputs = model.forward();

    // 检查输出
    if (outputs.empty()) {
        cerr << "Error: Model returned empty output." << endl;
        return;
    }

    // 确保输出维度匹配类别数
    if (outputs.total() != classNames.size()) {
        cerr << "Error: Model output dimension (" << outputs.total()
            << ") does not match number of classes (" << classNames.size() << ")" << endl;
        return;
    }

    // 获取预测结果
    Mat scores = outputs.reshape(1, 1);  // 确保是1行N列的矩阵
    Mat sortedIndices;
    sortIdx(scores, sortedIndices, SORT_EVERY_ROW + SORT_DESCENDING);

    auto end = chrono::high_resolution_clock::now();
    chrono::duration<double> elapsed = end - start;

    // 显示Top-5预测结果
    cout << "Top Predictions:" << endl;
    const int numToShow = min(5, static_cast<int>(classNames.size()));

    for (int i = 0; i < numToShow; ++i) {
        int idx = sortedIndices.at<int>(i);
        if (idx >= 0 && idx < classNames.size()) {
            printf("%d. %-20s: %.2f%%\n",
                i + 1,
                classNames[idx].c_str(),
                scores.at<float>(idx) * 100);
        }
    }

    cout << "\nInference time: " << elapsed.count() << " seconds" << endl;
}

int main() {
    const string modelPath = "D:/pythonPytorch/animaldemo/ResNet_best_model_20250514_201625.onnx";
    const string classNamesPath = "D:/pythonPytorch/animaldemo/name_of_the_animals.txt";
    const string imagePath = "D:/pythonPytorch/animaldemo/1.jpg";

    vector<string> classNames = loadClassNames(classNamesPath);
    if (classNames.size() != 90) {
        cerr << "Error: Must have exactly 90 classes" << endl;
        return -1;
    }

    Net net = readNetFromONNX(modelPath);
    if (net.empty()) {
        cerr << "Error: Failed to load model" << endl;
        return -1;
    }

    // 打印模型输入输出信息用于调试
    vector<String> layerNames = net.getLayerNames();
    cout << "Model layers: " << layerNames.size() << endl;

    predictImage(imagePath, net, classNames);

    return 0;
}

File: ResNet50_predict.py
import cv2
import numpy as np


def load_class_names(file_path):
    """
    从文件加载类别名称
    :param file_path: 类别名称文件路径
    :return: 类别名称列表
    """
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            class_names = [line.strip() for line in f.readlines()]
        return class_names
    except Exception as e:
        print(f"加载类别名称失败: {str(e)}")
        return None


def load_model(model_path):
    """
    加载ONNX模型
    :param model_path: 模型文件路径
    :return: OpenCV DNN模型
    """
    try:
        net = cv2.dnn.readNetFromONNX(model_path)
        if net.empty():
            raise ValueError("模型加载失败，模型为空。")
        return net
    except Exception as e:
        print(f"加载模型失败: {str(e)}")
        return None


def preprocess_image(image_path):
    """
    预处理输入图像，确保尺寸和通道匹配
    :param image_path: 图像路径
    :return: 预处理后的图像
    """
    try:
        img = cv2.imread(image_path)
        if img is None:
            raise ValueError(f"无法读取图像: {image_path}")

        # 确保图像大小为 224x224
        print(f"Original Image Shape: {img.shape}")  # 打印原始图像形状
        img_resized = cv2.resize(img, (224, 224))
        print(f"Resized Image Shape: {img_resized.shape}")  # 打印调整后的图像形状

        return img_resized
    except Exception as e:
        print(f"图像预处理失败: {str(e)}")
        return None


def predict_image(image_path, model, class_names):
    """
    使用OpenCV DNN模型对图像进行预测
    :param image_path: 输入图像路径
    :param model: 加载的OpenCV DNN模型
    :param class_names: 类别名称列表
    :return: 预测结果
    """
    img = preprocess_image(image_path)
    if img is None:
        return None

    try:
        # 创建 blob 对象 - 这里使用与PyTorch相同的预处理方式
        blob = cv2.dnn.blobFromImage(
            img,
            scalefactor=1.0 / 255.0,  # 归一化到[0,1]
            size=(224, 224),
            mean=[0.485, 0.456, 0.406],  # ImageNet均值
            swapRB=True,  # BGR转RGB
            crop=False
        )

        # 标准化
        blob[0, 0, :, :] = (blob[0, 0, :, :] - 0.485) / 0.229
        blob[0, 1, :, :] = (blob[0, 1, :, :] - 0.456) / 0.224
        blob[0, 2, :, :] = (blob[0, 2, :, :] - 0.406) / 0.225

        model.setInput(blob)
        outputs = model.forward()

        scores = outputs.flatten()
        top_indices = np.argsort(scores)[::-1][:5]
        top_scores = scores[top_indices]

        result = {
            'class_name': class_names[top_indices[0]],
            'confidence': top_scores[0],
            'all_scores': scores
        }
        return result
    except Exception as e:
        print(f"预测失败: {str(e)}")
        return None


if __name__ == "__main__":
    ONNX_MODEL_PATH = "ResNet_best_model_20250514_201625.onnx"
    CLASS_NAMES = load_class_names("name_of_the_animals.txt")

    if CLASS_NAMES is None:
        print("加载类别名称失败，程序终止")
    else:
        model = load_model(ONNX_MODEL_PATH)
        if model is None:
            print("加载模型失败，程序终止")
        else:
            result = predict_image("1.jpg", model, CLASS_NAMES)

            if result is None:
                print("预测失败")
            else:
                print(f"预测结果: {result['class_name']} (置信度: {result['confidence']:.2%})")
                print("\n所有类别置信度:")
                for i, (name, score) in enumerate(zip(CLASS_NAMES, result['all_scores'])):
                    if i < 5 or score > 0.1:
                        print(f"- {name}: {score:.2%}")